controllers/iamauthpolicy_controller.go

```diff
@@ -2,67 +2,80 @@ package controllers
 
 import (
 	"context"
-	"fmt"
+	"time"
 
 	anv1alpha1 "github.com/aws/aws-application-networking-k8s/pkg/apis/applicationnetworking/v1alpha1"
+	pkg_aws "github.com/aws/aws-application-networking-k8s/pkg/aws"
+	deploy "github.com/aws/aws-application-networking-k8s/pkg/deploy/lattice"
+	model "github.com/aws/aws-application-networking-k8s/pkg/model/lattice"
 	"github.com/aws/aws-application-networking-k8s/pkg/utils/gwlog"
 
-	k8serr "k8s.io/apimachinery/pkg/api/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
 type IAMAuthPolicyController struct {
-	log    gwlog.Logger
-	client client.Client
+	log       gwlog.Logger
+	client    client.Client
+	policyMgr deploy.IAMAuthPolicyManager
 }
 
-func RegisterIAMAuthPolicyController(log gwlog.Logger, mgr ctrl.Manager) error {
+func RegisterIAMAuthPolicyController(log gwlog.Logger, mgr ctrl.Manager, cloud pkg_aws.Cloud) error {
 	controller := &IAMAuthPolicyController{
 		log:    log,
 		client: mgr.GetClient(),
+		policyMgr: deploy.IAMAuthPolicyManager{
+			Cloud: cloud,
+		},
 	}
 	err := ctrl.NewControllerManagedBy(mgr).
 		For(&anv1alpha1.IAMAuthPolicy{}).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
 		Complete(controller)
 	return err
 }
 
 func (c *IAMAuthPolicyController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	policySpec := &anv1alpha1.IAMAuthPolicy{}
+	err := c.client.Get(ctx, req.NamespacedName, policySpec)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 	c.log.Infow("reconcile", "req", req)
 
-	policy := &anv1alpha1.IAMAuthPolicy{}
-	err := c.client.Get(ctx, req.NamespacedName, policy)
-	if !k8serr.IsNotFound(err) {
-		return ctrl.Result{}, err
+	policyModel, err := model.NewIAMAuthPolicy(policySpec)
+	if err != nil {
+		return ctrl.Result{RequeueAfter: time.Hour}, err // good candidate for terminal error
 	}
 
-	switch policy.Spec.TargetRef.Kind {
-	case "Gateway":
-		err = c.reconcileGateway(ctx, policy)
-		break
-	case "HTTPRoute":
-	case "GRPCRoute":
-		err = c.reconcileRoute(ctx, policy)
-		break
-	default:
-		err = fmt.Errorf("unsupported targetRef type, req=%s, kind=%s",
-			req, policy.Spec.TargetRef.Kind)
+	authPolicyFinalizer := "iamauthpolicy.k8s.aws/resources"
+
+	if policySpec.DeletionTimestamp.IsZero() {
+		policyStatus, err := c.policyMgr.Put(ctx, policyModel)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		policySpec.Annotations["ARN"] = policyStatus.Resource.ARN.String()
```

Comment:
It appears as though there is a convention in Kubernetes to set annotations with a prefix indicating the domain, such as what we use in ServiceImport: `multicluster.x-k8s.io`/<annotation-key>`.

> Automated system components (e.g. kube-scheduler, kube-controller-manager, kube-apiserver, kubectl, or other third-party automation) which add annotations to end-user objects must specify a prefix.

https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#:~:text=Syntax%20and%20character%20set

I'll make a change to fix this in AccessLogPolicy, since it's missed there too.

Reply:
make sense, will address in following pr. What's our approach on this? "aws.vpc.lattice.arn"?

Reply:
`vpc-lattice.amazonaws.com/arn` would follow the pattern, though this might warrant more discussion since it's a one-way door.

Also, regarding the suffix, `arn` doesn't explain what arn it is. Is this the arn of the resource on which the policy applies? If yes, the common name across AWS for this is `resourceArn`.

Reply:
will change ARN to ResourceId. `application-networking.k8s.aws/resourceId`